controllers/datamover.go

```diff
@@ -186,3 +203,176 @@ func (r *DPAReconciler) getDataMoverLabels() map[string]string {
+			if len(val) == 0 {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+func (r *DPAReconciler) createResticSecretsPerBSL(dpa *oadpv1alpha1.DataProtectionApplication, bsl velerov1.BackupStorageLocation, dmresticsecretname string, pass []byte) (*corev1.Secret, error) {
+
+	switch bsl.Spec.Provider {
+	case AWSProvider:
+		{
+			secretName, secretKey := r.getSecretNameAndKey(&bsl.Spec, oadpv1alpha1.DefaultPluginAWS)
+			bslSecret, err := r.getProviderSecret(secretName)
+			if err != nil {
+				return nil, err
+			}
+
+			awsProfile := "default"
+
+			if value, exists := bsl.Spec.Config[Profile]; exists {
+				awsProfile = value
+			}
+
+			key, secret, err := r.parseAWSSecret(bslSecret, secretKey, awsProfile)
+			if err != nil {
+				r.Log.Info(fmt.Sprintf("Error parsing provider secret %s for backupstoragelocation", secretName))
+				return nil, err
+			}
+			repobase := "s3:s3.amazonaws.com/"
+			if bsl.Spec.Config["s3Url"] != "" {
+				repobase = bsl.Spec.Config["s3Url"]
+			}
+
+			repo := repobase + bsl.Spec.ObjectStorage.Bucket
+			rsecret := &corev1.Secret{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      fmt.Sprintf("%s-volsync-restic", bsl.Name),
+					Namespace: bsl.Namespace,
+					Labels: map[string]string{
+						oadpv1alpha1.OadpBSLnameLabel:  bsl.Name,
+						oadpv1alpha1.OadpOperatorLabel: "True",
+					},
+				},
+			}
+
+			op, err := controllerutil.CreateOrUpdate(r.Context, r.Client, rsecret, func() error {
+
+				err := controllerutil.SetControllerReference(dpa, rsecret, r.Scheme)
+				if err != nil {
+					return err
+				}
+
+				return r.buildDataMoverResticSecret(rsecret, key, secret, bsl.Spec.Config[Region], pass, repo)
+			})
+
+			if err != nil {
+				return nil, err
+			}
+			if op == controllerutil.OperationResultCreated || op == controllerutil.OperationResultUpdated {
+				r.EventRecorder.Event(rsecret,
+					corev1.EventTypeNormal,
+					"ResticSecretReconciled",
+					fmt.Sprintf("%s restic secret %s", op, rsecret.Name),
+				)
+			}
+		}
+		// TODO: Azure & GCP
+	}
+
+	return nil, nil
+}
+
+//build data mover restic secret for given bsl
+func (r *DPAReconciler) buildDataMoverResticSecret(rsecret *corev1.Secret, key string, secret string, region string, pass []byte, repo string) error {
+
+	// TODO: add gcp, azure support
+	rData := &corev1.Secret{
+		Data: map[string][]byte{
+			AWSAccessKey:     []byte(key),
+			AWSSecretKey:     []byte(secret),
+			AWSDefaultRegion: []byte(region),
+			ResticPassword:   pass,
+			ResticRepository: []byte(repo),
+		},
+	}
+	rsecret.Data = rData.Data
+	return nil
+}
+
+func (r *DPAReconciler) ReconcileDataMoverResticSecret(log logr.Logger) (bool, error) {
+
+	// fetch latest DPA instance
+	dpa := oadpv1alpha1.DataProtectionApplication{}
+	if err := r.Get(r.Context, r.NamespacedName, &dpa); err != nil {
+		return false, err
+	}
+
+	// create secret only if data mover is enabled
+	if dpa.Spec.Features.DataMover != nil && dpa.Spec.Features.DataMover.Enable {
```

Comment:
Suggestion: Create a function which has all these checks on dpa and returns to you whether datamover is enabled or not, might make the code more readable and we can also escape from bugs if a dev forgets to adda nil check. WDTY ? @savitharaghunathan

Reply:
on it